main.py

```diff
@@ -0,0 +1,1254 @@
+    check_output_directory(config.get_output_dir())
+
+    # Check that output directory is writable.
+    if not os.access(config.get_output_dir(), os.W_OK):
+        logger.error(
+            "Output directory is not writable: %s", config.get_output_dir())
+        return 1
+
+    # Backup the output directory.
+    logger.info("Backing up current rules.")
+    backup_directory = util.mktempdir()
+    shutil.copytree(config.get_output_dir(), os.path.join(
+        backup_directory, "backup"), ignore=copytree_ignore_backup)
+
+    if not args.no_merge:
+        # The default, write out a merged file.
+        output_filename = os.path.join(
+            config.get_output_dir(), DEFAULT_OUTPUT_RULE_FILENAME)
+        file_tracker.add(output_filename)
+        write_merged(os.path.join(output_filename), rulemap)
+    else:
+        for filename in files:
+            file_tracker.add(
+                os.path.join(
+                    config.get_output_dir(), os.path.basename(filename)))
+        write_to_directory(config.get_output_dir(), files, rulemap)
+
+    if args.yaml_fragment:
+        file_tracker.add(args.yaml_fragment)
+        write_yaml_fragment(args.yaml_fragment, files)
+
+    if args.sid_msg_map:
+        write_sid_msg_map(args.sid_msg_map, rulemap, version=1)
+    if args.sid_msg_map_2:
+        write_sid_msg_map(args.sid_msg_map_2, rulemap, version=2)
+
+    if args.threshold_in and args.threshold_out:
+        file_tracker.add(args.threshold_out)
+        threshold_processor = ThresholdProcessor()
+        threshold_processor.process(
+            open(args.threshold_in), open(args.threshold_out, "w"), rulemap)
+
+    if not args.force and not file_tracker.any_modified():
+        logger.info("No changes detected, exiting.")
+        notes.dump_notes()
+        return 0
+
+    # Set these containers to None to fee the memory before testing Suricata which
+    # may consume a lot of memory by itself. Ideally we should refactor this large
+    # function into multiple methods so these go out of scope and get removed
+    # automatically.
+    rulemap = None
+    rules = None
+    files = None
+
+    if not test_suricata(suricata_path):
+        logger.error("Suricata test failed, aborting.")
+        logger.error("Restoring previous rules.")
+        copytree(
+            os.path.join(backup_directory, "backup"), config.get_output_dir())
+        return 1
+
+    if not config.args().no_reload and config.get("reload-command"):
+        logger.info("Running %s." % (config.get("reload-command")))
+        rc = subprocess.Popen(config.get("reload-command"), shell=True).wait()
+        if rc != 0:
+            logger.error("Reload command exited with error: %d", rc)
+
+    logger.info("Done.")
+
+    notes.dump_notes()
+
+    return 0
+
+def main():
+    try:
+        sys.exit(_main())
+    except exceptions.ApplicationError as err:
+        logger.error(err)
+    sys.exit(1)
+
+if __name__ == "__main__":
+    main()
```

Comment:
Wait. What happened here? Why is main.py a new file?

Reply:
Uh-oh. Seems like you've done a serious mess up with the paths. `main.py` is and should be in `suricata/update`

Reply:
I am unable to get a clear picture of what you're trying to do but what you're supposed to do is:
- Remove the matchers code from main.py exactly where it exists i.e. `suricata/update/main.py`.
Make a new file matchers.py in the same directory as main.py and move the matchers code there.

Reply:
Ma'am, I have done that only, so what should I upload the suricata/update folder? Or only matchers.py?

Reply:
What do you mean by uploading suricata/update folder?
And no, you've not done in this PR what I have mentioned above. You have instead created two new files `main.py` and `matchers.py` in the root dir of the project.

Reply:
Should I only upload matchers.py ? Because the code from Main.py is extracted that is why submitted that. I think by mistake I copied the path differently while creating the PR , should  , main.py and matchers.py that I have submitted our to put in suricata/update folder only ?So I 
submit matchers.py

Reply:
You should add the new matchers.py file in suricata/update folder and also add the modified main.py in suricata/update folder. Are you having trouble with git?

Reply:
Yes ma'am, maybe then I not submitting it the right way. Could you help me with if possible?

Reply:
Do the following from the root directory of your project. I hope you have a [remote](https://help.github.com/en/articles/adding-a-remote) pointing to this [current](https://github.com/OISF/suricata-update.git) git repository. I'm calling it `upstream` in the below guide.
```
git checkout master
git pull upstream master
```
This above step should not cause any merge conficts for you. If it does, make sure to save your work and do a `git reset --hard upstream master`.

```
git checkout -b matchers-2873-v2
cd suricata/update
```
Create a new file called `matchers.py`.
Remove matchers code from `main.py` and add it to `matchers.py`.
Save both the files.

`git status` should now show you that `main.py` was modified and `matchers.py` is a new _untracked_ file.

```
git add matchers.py main.py
git commit
```
Add the description about the changes you just made.

Check again if your changes in the files and your commit message correctly reflect by doing `git show`.

```
git push origin matchers-2873-v2
```
Create a PR then.